### PR TITLE
engram: host-side n-gram lookup for DeepSeek-V4.1, prefetched and overlapped

### PR DESCRIPTION
**Pull request by ZhangLirong-amd**
Refs #168, which prototyped the overlap on a standalone demo model.

engram: host-side n-gram lookup for DeepSeek-V4.1, prefetched and overlapped

DeepSeek-V4.1-Flash puts Engram modules on layers 1 and 14, backed by tables that cannot live in HBM: 384,006,168 and 384,016,682 rows of 256 fp8 values, about 98 GB each, 196B parameters in total. The lookup has to run on the host and it
has to be off the critical path, so this adds the host path, the device modules, and the runner hooks that hide the work behind the previous step's GPU time.

CPU precompute vs Engram recompute:
```
[atom.model_ops.engram_host 03:18:00] engram VERIFY VALUES step 1 seq=0 layer=1 token=32913 equal=True prefetched[:6]=[0.0703, -0.4375, 0.1875, -0.0049, -0.2344, -0.1719] recomputed[:6]=[0.0703, -0.4375, 0.1875, -0.0049, -0.2344, -0.1719]
[atom.model_ops.engram_host 03:18:00] engram VERIFY step 1: 0/2 rows mismatch, 0 wrong-token
[atom.model_ops.engram_host 03:18:00] engram VERIFY VALUES step 1 seq=0 layer=1 token=32913 equal=True prefetched[:6]=[0.0703, -0.4375, 0.1875, -0.0049, -0.2344, -0.1719] recomputed[:6]=[0.0703, -0.4375, 0.1875, -0.0049, -0.2344, -0.1719]
[atom.model_ops.engram_host 03:18:00] engram VERIFY step 1: 0/2 rows mismatch, 0 wrong-token
[atom.model_ops.engram_host 03:18:00] engram VERIFY VALUES step 1 seq=0 layer=1 token=32913 equal=True prefetched[:6]=[0.0703, -0.4375, 0.1875, -0.0049, -0.2344, -0.1719] recomputed[:6]=[0.0703, -0.4375, 0.1875, -0.0049, -0.2344, -0.1719]
[atom.model_ops.engram_host 03:18:00] engram VERIFY step 1: 0/2 rows mismatch, 0 wrong-token
[atom.model_ops.engram_host 03:18:00] engram VERIFY VALUES step 1 seq=0 layer=1 token=32913 equal=True prefetched[:6]=[0.0703, -0.4375, 0.1875, -0.0049, -0.2344, -0.1719] recomputed[:6]=[0.0703, -0.4375, 0.1875, -0.0049, -0.2344, -0.1719]
[atom.model_ops.engram_host 03:18:00] engram VERIFY step 1: 0/2 rows mismatch, 0 wrong-token

```

| | Paper demo (`engram_demo_v1.py`) | DeepSeek-V4.1-Flash |
|---|---|---|
| **Rows per hash head** | 129,280 × 5 ≈ 646k | 16,000,000 |
| **N-gram orders** | max 3 → {2, 3} → 16 heads | max 4 → {2, 3, 4} → 24 heads |
| **Embedding dim per head** | 512 / 8 = 64 | 256 |
| **Rows per layer** | ≈ 10.3M | 384,006,168 (layer 1) / 384,016,682 (layer 14) |
| **Table size per layer** | ≈ 2.6 GB (fp32) | 98.3 GB (fp8 e4m3 + e8m0 block scales) |
| **Total engram parameters** | ≈ 1.3B | 196B across both modules |
| **Layer placement** | [1, 15] | [1, 14] |
| **Short causal conv** | present | omitted (tech report §2.4.2) |
| **Key/value projections** | 4 separate `Linear` + 1 `Linear` | one fused `wkv` `[25600, 6144]` |
| **Quantization** | none (fp32 `nn.Embedding`) | fp8 throughout, 32-value blocks on the table, 32×32 on `wkv` |

Engram host-overlap  (DeepSeek-V4.1)
```
 step N      ┌─ model_runner ─────────────────────────────────────────┐
 (overlap    │  sampled token ─► CPU ─► hash ─► gather embedding       │
  postproc)  │                            └─ staging buffer ─► H2D ───┐ │
             └─────────────────────────────────────────────────────── │─┘
                                                                       ▼
 step N+1    ┌─ Engram op  (layers 1 & 14) ─────────────────────────────┐
 (forward)   │  load staged embedding ─► wkv + gate·value ─► += hidden   │
             └───────────────────────────────────────────────────────── ┘
```

image

### CODE REVIEW

File: atom/model_ops/engram.py [hunk 1]
```diff
@@ -0,0 +1,1162 @@
...
+        # the real table that is ~0.9 us/row against ~8 us/row for a Python loop.
+        rows = self._tensor[index].to(out_dtype)
+        if self._scale is not None:
+            # Block-quantized: each scale covers `block_size` consecutive values
+            # of a row. Skipping this does not fail, it returns values two orders
+            # of magnitude off, so it is not optional.
+            scale = decode_block_scale(self._scale[index], out_dtype)
+            rows = (
+                rows.reshape(-1, scale.shape[1], self.block_size) * scale.unsqueeze(-1)
+            ).reshape(-1, self.head_dim)
```
[the remark this answers is not in the document]
Comment (ZhangLirong-amd) [reply atom/model_ops/engram.py:527]: gather() flattens row_indices first (`flat = row_indices.reshape(-1)`; `index = torch.from_numpy(flat)` is 1-D), so `self._scale[index]` is always `[num_gathered, n_blocks]` regardless of the leading dims of row_indices — `scale.shape[1]` is therefore the block count, not a leading dim. The trailing `reshape(*row_indices.shape, head_dim)` restores the original shape. Added test_host_table_gather_scale_multi_leading_dims (a [B, heads] scaled gather) to lock this in.

File: atom/model_ops/engram_layer.py [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+import os
+
+import torch
+from torch import nn
+
+from atom.model_ops.engram import (
+    CompressedTokenizer,
+    EngramConfig,
+    EngramHost,
+    EngramPrefetcher,
```
[the remark this answers is not in the document]
Comment (ZhangLirong-amd) [reply atom/model_ops/engram_layer.py:43]: EngramPrefetcher is used in build_engram_host (engram_layer.py, the `EngramPrefetcher(self.hash_mapping, self._tables, max_concurrent_seqs=max_num_tokens)` call), so it is not an unused import — `ruff check` passes clean on this file. Resolving as a false positive.

File: recipes/Engram.md [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+Engram (DeepSeek's https://github.com/deepseek-ai/Engram) augments a few decoder
+layers with a lookup into very large n-gram embedding tables. The tables cannot
+live in HBM (DeepSeek-V4.1-Flash carries two, ~101.5 GB each), so the lookup runs
+on the **host** and its result is staged to the device.
+
+This recipe documents how a **model** wires Engram in. The engram code is a
+self-contained library; no in-tree model uses it today. The DeepSeek-V4.1 stub we
+used to bring it up and verify correctness has been removed — it lives in git
+history (`git show 00afbe32b`). If a model's config declares engram layers but the
+model does not expose `build_engram_host`, the runner **fails closed** (it raises
```
[the remark this answers is not in the document]
Comment (ZhangLirong-amd) [reply recipes/Engram.md:12]: Narrowed the PR to match: the title is now "engram: host-side n-gram embedding lookup — library + runner overlap contract" and the description leads with a Scope note stating this lands the library + runner contract (not an in-tree DeepSeek-V4.1 model), that no shipped model wires Engram today, that the runner fails closed when a config declares engram but the model lacks build_engram_host, and that the V4.1 bring-up stub lives in git history (git show 00afbe32b). recipes/Engram.md already documents it as the model-side contract with no in-tree consumer. Wiring Engram into a shipped model is follow-up work.

**Comment by ZhangLirong-amd** [comment]
close for #2314 finished merged
